pkg/querier/frontend/worker_frontend_manager.go

```diff
@@ -0,0 +1,183 @@
+package frontend
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/weaveworks/common/httpgrpc"
+	"github.com/weaveworks/common/httpgrpc/server"
+	"go.uber.org/atomic"
+
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+var (
+	backoffConfig = util.BackoffConfig{
+		MinBackoff: 50 * time.Millisecond,
+		MaxBackoff: 1 * time.Second,
+	}
+
+	errGracefulQuit = errors.New("processor quitting gracefully")
+)
+
+type frontendManager struct {
+	client FrontendClient
+	server *server.Server
+
+	log            log.Logger
+	maxSendMsgSize int
+
+	gracefulQuit      []chan struct{}
```

Comment:
My gut feeling here is we need a map of cancel funcs, not close channels - one per worker goroutine.  That will allow us to interrupt workers to scale down and stop more gracefully.

Reply:
Swapped to context.CancelFunc's.  the frontend will now force cancel queries on scaling.  There's pros/cons here, but ultimately I am agreeing with this approach.